### ISSUE TITLE
Functions that call `_removeSession` internally do not trigger `SIGNED_OUT` event when the function fails

### ISSUE THREAD

**Issue by bombillazo** [supabase/gotrue-js#853]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

When using the client and auth functions, some internally call a `_removeSession` function, presumably to clear the session before attempting any auth-related logic. This effectively clears out the local session data used to keep the user logged in. In many cases the function runs and restores or resets the session data eventually. However, if the function fails, this does not trigger the SIGNED_OUT to notify any even handlers and hooks that the session data is no longer present, causing inconsistent auth data states.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Create a supabase client instance.
2. In your app, use the `auth-helpers-react` lib and use the `useSession` hook to get the session data.
3. Trigger a function like `supabase.auth.resend` with invalid params to trigger an error and session remove.
4. This causes the session data to be erased, but the context data of the `useSession` hook is not updated and retains the old values.

## Expected behavior

Any auth function that fails and deletes the auth session data should trigger a `SIGNED_OUT` event to notify any other serives.

## System information

- Version of supabase-js: 2.39.6

**Comment by bombillazo** [comment]
Any update on this issue?

**Comment by kangmingtay** [comment]
@bombillazo hmm there shouldn't be a session return by `useSession` when the `supabase.auth.resend` method is called because no user is actually signed in yet

**Comment by bombillazo** [comment]
We can call resend to change email or phone. I dont expect users to be signout out for this

**Comment by kangmingtay** [comment]
@bombillazo yeah we handle that [case](https://github.com/supabase/auth-js/blob/4ecfdda65188b71322753e57622be8eafe97ed6b/src/GoTrueClient.ts#L877-L879) already by not removing the session if the type passed indicates an email change or phone change

**Comment by bombillazo** [comment]
This issue is more related to what happens when something fails with the `auth` calls while using the `useSession` react hook that returns the session. Since this library isn't properly triggering the `SIGNED_OUT` event, when an auth call fails and erases the session from the local client, the react hook keeps returning the cached session.

**Comment by bombillazo** [comment]
This is still open and should be fixed with #854
